End to end hack to call server side Caffe2 ops

dzhulgakov opened the pull request:
Summary:
Motivation: we don't actually want to use it for real under any circumstances. This is an idea to unblock our internal progress and parallelize workstreams. We can easily define schemas for all ops in question and implement forwarding to C2 ops which is NOT going to be performant. Then several things can be happening in parallel:
* move code of ops outside of C2 ops that depend on protobuf into c10
* development of optimization/fusion passes
* building python-level wrappers with clean API
* improving perf

This demonstrates, Relu, quant, dequant. It seems to cover all use cases necessary (maybe except weights prepacking). Ideally I'd demonstrate Conv, but will get to it later in a separate PR (contributions welcomed)

Differential Revision: D14531232

ezyang commented on torch/csrc/jit/register_quantized_ops.cpp:
```diff
@@ -0,0 +1,146 @@
+namespace {
+
+caffe2::Tensor from_at_tensor(c10::IValue v) {
+  return caffe2::Tensor(autograd::Variable(std::move(v).toTensor()).data());
```
Eek! Do you /really/ want a `caffe2::Tensor` that's a Variable? This feels incredibly unlikely to be what you actually want.

dzhulgakov replied:
There's .data() in the end - so it gets the underlying tensor. In fact Caffe2 tensor has an enforce against variables

ezyang commented on torch/csrc/jit/register_quantized_ops.cpp:
```diff
@@ -0,0 +1,146 @@
+        // NOTE: we put outout in double parens because it's an output of type
+        // tuple, not a tuple of multiple outputs
+        "c10::quantized_relu((Tensor, float, int) self) -> ((Tensor, float, int))",
+        // TODO: can't use C++ inference - doesn't work yet for tuple types
```
cc @smessmer

ezyang commented on torch/csrc/jit/register_quantized_ops.cpp:
```diff
@@ -0,0 +1,146 @@
+
+          // TODO: refactor the underlying op implementation and inline it in
+          // c10 kernel
+          caffe2::Workspace ws;
```
speeeeshul :3

ezyang reviewed (approved):
Fine for unblock.

Commit:
End to end hack to call server side Caffe2 ops (#18267)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/18267

Motivation: we don't actually want to use it for real under any circumstances. This is an idea to unblock our internal progress and parallelize workstreams. We can easily define schemas for all ops in question and implement forwarding to C2 ops which is NOT going to be performant. Then several things can be happening in parallel:
* move code of ops outside of C2 ops that depend on protobuf into c10
* development of optimization/fusion passes
* building python-level wrappers with clean API
* improving perf

This demonstrates, Relu, quant, dequant. It seems to cover all use cases necessary (maybe except weights prepacking). Ideally I'd demonstrate Conv, but will get to it later in a separate PR (contributions welcomed)

Reviewed By: ezyang

Differential Revision: D14531232

fbshipit-source-id: 68e2d4be407b95fc01f2494d0485c9eb92922c65
